src/Rules/Program.hs

```diff
@@ -1,116 +1,100 @@
-        let context0 = vanillaContext Stage0 package -- TODO: get rid of context0
-        inplaceBinPath -/- programName context0 <.> exe %> \bin -> do
-            stage <- installStage package -- TODO: get rid of fromJust
-            buildBinaryAndWrapper rs bin =<< programContext (fromJust stage) package
-
-        inplaceLibBinPath -/- programName context0 <.> exe %> \bin -> do
-            stage   <- installStage package -- TODO: get rid of fromJust
-            context <- programContext (fromJust stage) package
-            if package /= iservBin then
-                -- We *normally* build only unwrapped binaries in inplace/lib/bin
-                buildBinary rs bin context
-            else
-                -- Build both binary and wrapper in inplace/lib/bin for iservBin
-                buildBinaryAndWrapperLib rs bin context
-
-        inplaceLibBinPath -/- programName context0 <.> "bin" %> \bin -> do
-            stage <- installStage package -- TODO: get rid of fromJust
-            buildBinary rs bin =<< programContext (fromJust stage) package
-
-buildBinaryAndWrapperLib :: [(Resource, Int)] -> FilePath -> Context -> Action ()
-buildBinaryAndWrapperLib rs bin context = do
-    windows <- windowsHost
-    if windows
-    then buildBinary rs bin context -- We don't build wrappers on Windows
-    else case lookup context inplaceWrappers of
-        Nothing      -> buildBinary rs bin context -- No wrapper found
-        Just wrapper -> do
-            top <- topDirectory
-            let libdir = top -/- inplaceLibPath
-            let wrappedBin = inplaceLibBinPath -/- programName context <.> "bin"
-            need [wrappedBin]
-            buildWrapper context wrapper bin (WrappedBinary libdir (takeFileName wrappedBin))
-
-buildBinaryAndWrapper :: [(Resource, Int)] -> FilePath -> Context -> Action ()
-buildBinaryAndWrapper rs bin context = do
-    windows <- windowsHost
-    if windows
-    then buildBinary rs bin context -- We don't build wrappers on Windows
-    else case lookup context inplaceWrappers of
-        Nothing      -> buildBinary rs bin context -- No wrapper found
-        Just wrapper -> do
-            top <- topDirectory
-            let libPath    = top -/- inplaceLibPath
-                wrappedBin = inplaceLibBinPath -/- takeFileName bin
-            need [wrappedBin]
-            buildWrapper context wrapper bin (WrappedBinary libPath (takeFileName bin))
-
-buildWrapper :: Context -> Wrapper -> FilePath -> WrappedBinary -> Action ()
-buildWrapper context@Context {..} wrapper wrapperPath wrapped = do
-    contents <- interpretInContext context $ wrapper wrapped
-    writeFileChanged wrapperPath contents
-    makeExecutable wrapperPath
-    putSuccess $ "| Successfully created wrapper for " ++
-        quote (pkgName package) ++ " (" ++ show stage ++ ")."
+              cross <- crossCompiling
+              -- for cross compiler. copy the stage0/bin/<pgm>
+              -- into stage1/bin/
```

Comment:
So, for cross compilation, we don't really want the stage2 compiler (when host /= target). We want the compiler that runs on the the host (e.g. the one we compile the compiler on) but targets what ever we want to target.

To make the bin-dist logic simpler. And end up with the right `ghc` executable in the final `bin` directory, for cross compilers we take the stage0 binary (used to be called `ghc-stage1`) and copy it into stage1 (used to be called `ghc-stage2`). Similarly for other binaries, but only when building a cross compiler.

This might be at odds when we want to build a cross compiled ghc, as opposed to ghc as a cross compiler.

Reply:
In that case we would *not* want this logic, and use the stage1 compiler (build == host -> target) to produce the stage2 compiler (host == target /= build)

Reply:
"compiler. copy" --> "compiler, copy"

Reply:
@ggreif, you are one hell of a proofreader! 💯